File: Algorithm/Sort & Order Statistic/RanSelect.m
% 期望为线性时间的顺序统计量查找算法，p,r为数组下标，求取第i个顺序统计量num
function num=RanSelect(Array,p,r,i)
if p==r
    num=Array(p);
    return;
end
[Array,q]=sub_ranPartition(Array,p,r);
k=q-p+1;
if i==k
    num=Array(q);
elseif i<k
    num=RanSelect(Array,p,q-1,i);
else
    num=RanSelect(Array,q+1,r,i-k);
end
end

function [A,q]=sub_partition(A,p,r)
x=A(r);
i=p-1;
for j=p:r-1
    if A(j)<=x
        i=i+1;
        temp=A(i);
        A(i)=A(j);
        A(j)=temp;
    end
end
temp=A(i+1);
A(i+1)=A(r);
A(r)=temp;
q=i+1;
end

function [A,q]=sub_ranPartition(A,p,r)
i=round(p+(r-p)*rand());
temp=A(i);
A(i)=A(r);
A(r)=temp;
[A,q]=sub_partition(A,p,r);
end

File: Algorithm/Sort & Order Statistic/C/RanSelect.c
//添加stdlib.h及time.h
int sub_partition(int* A, int p, int r)
{
	int x = *(A + r - 1);
	int j, i = p - 2;
	int temp;
	for (j = p - 1; j < r - 1; j++)
		if (*(A + j) <= x)
		{
			i++;
			temp = *(A + i);
			*(A + i) = *(A + j);
			*(A + j) = temp;
		}
	temp = *(A + i + 1);
	*(A + i + 1) = *(A + r - 1);
	*(A + r - 1) = temp;
	int q = i + 1;
	return q;
}

int sub_ranPartition(int* A, int p, int r)
{
	srand((unsigned)time(NULL));    //产生伪随机数序列的种子，应放在主程序中
	int temp;
	int i = rand() % (r - p + 1) + p;
	temp = *(A + i - 1);
	*(A + i - 1) = *(A + r - 1);
	*(A + r - 1) = temp;
	int q = sub_partition(A, p, r);
	return q;
}

int RanSelect(int* A, int p, int r, int i)
{
	int num;
	if (p == r)
	{
		num = *(A + p - 1);
		return num;
	}
	int q = sub_ranPartition(A, p, r) + 1;
	int k = q - p + 1;

	if (i == k)
		num = *(A + q - 1);
	else if (i < k)
		num = RanSelect(A, p, q - 1, i);
	else
		num = RanSelect(A, q + 1, r, i - k);
	return num;
}